### ISSUE TITLE
Incorrect SSA value type when using `julia --code-coverage`

### ISSUE THREAD

**Issue by KDr2** [JuliaCompilerPlugins/CodeInfoTools.jl#18]
I have a `t.jl`:

```
using CodeInfoTools

struct S x; y; end

c = CodeInfoTools.code_inferred(S, Int, Int)

@show c
@show c.code
@show c.slottypes
@show c.ssavaluetypes
```

When I run `julia t.jl`, it produces the correct output:
```
c = CodeInfo(
    @ /data4/zhuo/Work/julia/Libtask.jl/test/runtests.jl:3 within `S`
1 ─ %1 = Main.S::Core.Const(S)
│   %2 = %new(%1, x, y)::S
└──      return %2
)
c.code = Any[:(Main.S), :(%new(%1, _2, _3)), :(return %2)]
c.slottypes = Any[Core.Const(S), Int64, Int64]
c.ssavaluetypes = Any[Core.Const(S), S, Any]
```

But when I run `julia --code-coverage=user t.jl`, I get:
```
c = CodeInfo(
    @ /data4/zhuo/Work/julia/Libtask.jl/test/runtests.jl:3 within `S`
1 ─ %1 = Main.S::Nothing
│   %2 = %new(%1, x, y)::Core.Const(S)
└──      return %2
)
c.code = Any[:(Main.S), :(%new(%1, _2, _3)), :(return %2)]
c.slottypes = Any[Core.Const(S), Int64, Int64]
c.ssavaluetypes = Any[Nothing, Core.Const(S), S, Any]
```
There's an extra `Nothing` prepended to the `ssavalutypes`.

It's mainly caused by the `--code-coverage` option:

- correct: julia  t.jl
- correct: julia --code-coverage=none t.jl
- incorrect: julia --code-coverage=user t.jl
- incorrect: julia --code-coverage=all t.jl

**Comment by KDr2** [comment]
The real reason is not an extra `Nothing` is prepended, but an expression is removed:

```
julia> code_typed(S, Tuple{Int, Int})
1-element Vector{Any}:
 CodeInfo(
1 ─      $(Expr(:code_coverage_effect))::Nothing # <-- this expr is removed, but its type isn't
│   %2 = Main.S::Type{S}
│   %3 = %new(%2, x, y)::S
└──      return %3
) => S
```

**Comment by femtomc** [comment]
@KDr2 did you end up figuring a workaround for this?

I can't tell how much of this is an issue with `CodeInfoTools` -- vs. the fact that Julia will insert these "marker" expressions for coverage.

**Comment by KDr2** [comment]
My workaround is not to use the `--code-coverage` option so far.  :)

**Comment by Roger-luo** [comment]
FYI: https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl/blob/master/src/utils.jl#L82

you can just use this transform on the generated `CodeInfo`

**Comment by KDr2** [comment]
> FYI: https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl/blob/master/src/utils.jl#L82
> 
> you can just use this transform on the generated `CodeInfo`

Thank you @Roger-luo, but it has no effect on this issue:

```
$ cat t.jl
using CodeInfoTools
using CompilerPluginTools

struct S x; y; end

c = CodeInfoTools.code_inferred(S, Int, Int)
c = CompilerPluginTools.rm_code_coverage_effect(c)

@show c
@show c.code
@show c.slottypes
@show c.ssavaluetypes


$ julia --project t.jl
c = CodeInfo(
    @ /data4/zhuo/Work/julia/Libtask.jl/t.jl:4 within `S`
1 ─ %1 = Main.S::Const(S)
│   %2 = %new(%1, x, y)::S
└──      return %2
)
c.code = Any[:(Main.S), :(%new(%1, _2, _3)), :(return %2)]
c.slottypes = Any[Const(S), Int64, Int64]
c.ssavaluetypes = Any[Const(S), S, Any] # <--- HERE is correct

$ julia --code-coverage --project t.jl
c = CodeInfo(
    @ /data4/zhuo/Work/julia/Libtask.jl/t.jl:4 within `S`
1 ─ %1 = Main.S::Nothing
│   %2 = %new(%1, x, y)::Const(S)
└──      return %2
)
c.code = Any[:(Main.S), :(%new(%1, _2, _3)), :(return %2)]
c.slottypes = Any[Const(S), Int64, Int64]
c.ssavaluetypes = Any[Nothing, Const(S), S]  # <--- HERE still a prepended Nothing
```

**Comment by Roger-luo** [comment]
I think you don't want to use CodeInferred since that contains the type lattice object. That pass should not be run with type lattice as the name suggested it removes the coverage effect stmts.  So you should just use it on the final codeinfo. I'm not sure why there's code inferred in this package but the CodeInfo manipulation should be better written in IRCode and we should only use CodeInfo as a final product after them. Unless you have a custom abstract interpreter.

**Comment by KDr2** [comment]
But I need the type info of the `CodeInfo` object. And `rm_code_coverage_effect` doesn't indicate anything about types, i.e., it doesn't say it can/can't deal with types, so I think if a `CodeInfo` contains IR codes and types, it may remove coverage related things from both of them.

Anyway, I found `Base.code_typed` returns the same code info (with types and without errors), so I changed to that function.

Thank you.
